src/Microsoft.VisualStudio.SDK.Analyzers/VSSDK004ProvideAutoLoadAttributeAnalyzer.cs

```diff
@@ -0,0 +1,83 @@
+            isEnabledByDefault: true);
+
+        /// <inheritdoc />
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Descriptor);
+
+        /// <inheritdoc />
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze);
+
+            // Register for compilation first so that we only activate the analyzer for applicable compilations
+            context.RegisterCompilationStartAction(compilationContext =>
+            {
+                var asyncPackage = compilationContext.Compilation.GetTypeByMetadataName(Types.AsyncPackage.FullName);
+                if (asyncPackage != null)
+                {
+                    var autoLoadAttribute = compilationContext.Compilation.GetTypeByMetadataName(Types.ProvideAutoLoadAttribute.FullName);
+                    var packageAutoLoadFlags = compilationContext.Compilation.GetTypeByMetadataName(Types.PackageAutoLoadFlags.FullName);
+
+                    // Reuse the type symbols we looked up so that we don't have to look them up for every single class declaration.
+                    compilationContext.RegisterSyntaxNodeAction(Utils.DebuggableWrapper(ctxt => this.AnalyzeClassDeclaration(ctxt, autoLoadAttribute, packageAutoLoadFlags)), SyntaxKind.ClassDeclaration);
+                }
+            });
+        }
+
+        private void AnalyzeClassDeclaration(SyntaxNodeAnalysisContext context, INamedTypeSymbol autoLoadAttributeType, INamedTypeSymbol packageAutoLoadFlagsType)
+        {
+            var declaration = (ClassDeclarationSyntax)context.Node;
+            var userClassSymbol = context.SemanticModel.GetDeclaredSymbol(declaration, context.CancellationToken);
+
+            // Enumerate all attribute lists and attribute to find ProvideAutoLoad attributes as there can be multiple ones
+            foreach (var autoLoadInstance in userClassSymbol?.GetAttributes().Where(a => a.AttributeClass == autoLoadAttributeType))
+            {
+                var flagsArgument = autoLoadInstance.ConstructorArguments.FirstOrDefault(p => p.Type == packageAutoLoadFlagsType);
+                Shell.PackageAutoLoadFlags flagsValue = flagsArgument.IsNull ? PackageAutoLoadFlags.None : (Shell.PackageAutoLoadFlags)flagsArgument.Value;
+
+                // We are looking for any auto load attribute without the BackgroundLoad flag and SkipWhenUIContextRulesActive flag.
+                if (!flagsValue.HasFlag(Shell.PackageAutoLoadFlags.BackgroundLoad) &&
+                    !flagsValue.HasFlag(Shell.PackageAutoLoadFlags.SkipWhenUIContextRulesActive))
+                {
+                    var attributeSyntax = autoLoadInstance.ApplicationSyntaxReference.GetSyntax(context.CancellationToken) as AttributeSyntax;
+                    Location location = attributeSyntax.GetLocation();
```

Comment:
You're dereferencing `attributeSyntax` after using the `as` cast operator, which may return null. See https://aka.ms/ascast for my blog post on this. In this case, since in some cases `GetSyntax` may perhaps return null (say, if the attribute were inherited from a base class that is defined in a referenced assembly) I think you should be prepared for null here.